### PR TITLE
Add settings history.defaultPageSize

### LINKED ISSUE #44431: Create a configurable where admins can set a default page limit on the file history commits, and users can configure their own.
- **Requested by:** https://github.com/sourcegraph/accounts/issues/6716 via [Slack](https://sourcegraph.slack.com/archives/C01D50MSA7L/p1668535192009239)

#### Feature request description

When viewing the history of a file, how many commits are loaded in each page of commits, before a Show more link is shown at the bottom? Can we create a configurable where admins can set a default to ~5, and users can change that if they like?

#### Is your feature request related to a problem? If so, please describe.

For large repositories, it can take a significant amount of time to load this history tab and most users may not need to see 100 commits.

/cc @jplahn @ryphil

### PR DESCRIPTION
**Pull request by indradhanush**
Fixed #44431.

https://user-images.githubusercontent.com/2682729/203026948-93dd8be7-b1ca-4172-834a-164fa3f1f62d.mp4

**Note about returning 2x the responses on clicking the `Show more` button:**

- This is the existing behaviour and unchanged
- Initially we load say 5 commits
- Next time it fetched 10 commits, so visually it looks like 5 commits were fetched (the API doesn't support pagination yet)
- Then it fetches 20 commits which is why visually it looks like 10 new commits were fetched and thus the doubling of the commits from this point on

## Test plan

Tested functionality locally.
Builds (should) pass.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.00% (+0.38 kb) | 0.00% (+0.38 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/ig/custom-history-page-size/compare-report.html;tab=live_object) for a full comparison between the commits 27fb2bb and 6bdfb4b or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Comment by ryanslade** [comment]
The doubling up behaviour is weird. In your example of fetching 5 commits I would expect it to fetch 5, 10, 15, 20 etc commits each time you click "show more" so that it *looks* like it's loading 5 new commits each time. 

It looks like the increments themselves are going up which is a weird user experience. It was easy to spot near the end of your demo where you set the page size to 2. Is this the existing behaviour you mentioned?

**Comment by indradhanush** [comment]
@ryanslade Right. Ideally it should be in increments of `5` if the page size is set to `5`. But the current behaviour on dotcom says otherwise. The page size is `100` and by the third request we fetch `400` instead of `300` commits.

image

The existing behaviour is not changing but with a smaller page size, this is more pronounced.

File: schema/settings.schema.json [hunk 1]
```diff
@@ -736,6 +736,12 @@
       "type": "boolean",
       "default": false
     },
+    "history.defaultPageSize": {
+      "description": "Custom page size for the history tab. If set, the history tab will populate that number of commits at a time. If not set, it will populate 100 commits at a time.",
+      "type": "integer",
+      "default": 100,
```
Comment (mrnugget) [inline schema/settings.schema.json:742]: If you set a default here then that will be used and the value is never `undefined` in JS, correct? Does it make sense to set a default here? Or should we not set one in JS if we have it here?

Comment (indradhanush) [reply schema/settings.schema.json:742]: That's a good point. I think to ease of the confusion of having a `default` value for a value that is named `defaultPageSize`, maybe we should remove the default from the schema config.

**Comment by ryanslade** [comment]
> The existing behaviour is not changing but with a smaller page size, this is more pronounced.

I wonder if we should change this? It's the sort of thing that will probably come back to us as a bug report.

**Review by mrnugget — APPROVED** [review]

**Comment by mrnugget** [comment]
> I wonder if we should change this? It's the sort of thing that will probably come back to us as a bug report.

➕ on this. But I'd vote to have this in a separate PR. There's also efforts going on to improve pagination (cc @thenamankumar) so let's make sure to not step on each other's toes.

**Review by thenamankumar — APPROVED** [review]

**Comment by thenamankumar** [comment]
@ryanslade @mrnugget I think the current behaviour was implemented as a UX improvement which might not look like one to us at the moment. But we should discuss this with the design team first before planning on removing it as the FilteredConnection component, which implements this logic, is used at multiple places.

**Comment by mrnugget** [comment]
@thenamankumar thanks for the reminder. Yes, agree. I thought it was a legitimate bug, but just watched the video and agree, this seems like the intended behaviour, even though it's weird.

**Pull request merged**